logger.py

```diff
@@ -22,6 +22,14 @@ def error_and_exit(self: 'ExtendedLogger', msg: str, *, exit_code: int = -1) ->
         self.error(msg)
         sys.exit(exit_code)
 
+    def info_once(self, last_msg: None | str, msg: str) -> str:
```

Comment:
`None | str` is a red flag.

Reply:
it's functionally the same as `Optional[str]`. Changed to `Optional[str]`.